tools/cfn2ts/lib/codegen.ts

```diff
@@ -514,15 +520,26 @@ export default class CodeGenerator {
     this.code.closeBlock();
   }
 
-  private emitProperty(context: genspec.CodeName, propName: string, spec: schema.Property, additionalDocs: string): string {
-    const question = spec.Required ? '' : '?';
-    const javascriptPropertyName = genspec.cloudFormationToScriptName(propName);
+  private emitProperty(props: EmitPropertyProps): string {
+    const question = props.spec.Required ? '' : '?';
+    const javascriptPropertyName = genspec.cloudFormationToScriptName(props.propName);
 
-    this.docLink(spec.Documentation, additionalDocs);
-    this.code.line(`readonly ${javascriptPropertyName}${question}: ${this.findNativeType(context, spec, propName)};`);
+    this.docLink(props.spec.Documentation, props.additionalDocs);
+    const line = `: ${this.findNativeType(props.context, props.spec, props.propName)};`;
 
+    if (props.container === Container.Interface) {
+        this.code.line(`readonly ${javascriptPropertyName}${question}${line}`);
+    }
+    if (props.container === Container.Class) {
+      if (props.propName === 'Tags' && schema.isTagProperty(props.spec)) {
+        this.code.line(`public readonly tags: ${TAG_MANAGER};`);
+      } else {
+      this.code.line(`public ${javascriptPropertyName}Prop${line}`);
```

Comment:
Can you share some examples of conflicts that required adding the `Prop` suffix. I actually think the prop suffix is not that bad, but I am a bit surprised because I was under the assumption that attributes all start with the type name (e.g. `topicName` and props usually don't). But obvsiouly I am wrong.

Reply:
@eladb -- great follow-up because I didn't look close enough. The collision that I'm seeing with IAM module is actually caused by our `Ref` implementation.

```ts
export class CfnUser extends cdk.CfnResource {
    /**
     * The CloudFormation resource type name for this resource class.
     */
    public static readonly resourceTypeName = "AWS::IAM::User";

    /**
     * @cloudformationAttribute Arn
     */
    public readonly userArn: string;
    /* this is the most common collision */
    public readonly userName: string;

    /**
     * `AWS::IAM::User.Groups`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html#cfn-iam-user-groups
     */
    public groups: string[];

    /**
     * `AWS::IAM::User.LoginProfile`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html#cfn-iam-user-loginprofile
     */
    public loginProfile: CfnUser.LoginProfileProperty | cdk.Token;

    /**
     * `AWS::IAM::User.ManagedPolicyArns`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html#cfn-iam-user-managepolicyarns
     */
    public managedPolicyArns: string[];

    /**
     * `AWS::IAM::User.Path`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html#cfn-iam-user-path
     */
    public path: string;

    /**
     * `AWS::IAM::User.PermissionsBoundary`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html#cfn-iam-user-permissionsboundary
     */
    public permissionsBoundary: string;

    /**
     * `AWS::IAM::User.Policies`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html#cfn-iam-user-policies
     */
    public policies: Array<CfnUser.PolicyProperty | cdk.Token> | cdk.Token;

    /**
     * `AWS::IAM::User.UserName`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html#cfn-iam-user-username
     */
    public userName: string;

    /**
     * Create a new `AWS::IAM::User`.
     *
     * @param scope - scope in which this resource is defined
     * @param id    - scoped id of the resource
     * @param props - resource properties
     */
    constructor(scope: cdk.Construct, id: string, props?: CfnUserProps) {
        super(scope, id, { type: CfnUser.resourceTypeName, properties: props });
        this.userArn = this.getAtt('Arn').toString();
        this.userName = this.ref.toString();
    }
 /* ... snip ... */
```

So the issue is actually the `Ref` that has the same name as a property. Would you prefer we suffix these special cases?

Reply:
I actually wanted to completely get rid of these RefKinds (see #2345). Happy to include this in this CR if you are inclined, although it will have a pretty substantial breaking effect on the entire framework (not too hard to fix, but annoying).

Reply:
Can you give me an example of what eliminating RefKinds would look like in the above example? #2345 leads me to believe also eliminating the patching of the spec?

Reply:
Yes, basically remove the patching of the spec and the logic that generates bucketName in the above example. See my general comment below.